File: frontend-new/libs/shared/design-system/src/api/modules/conformance.ts
/**
 * Conformance Module - SDK methods for conformance checking
 * 
 * FIXED: Standardized on datasetId (was datasetId)
 */

import type { ApiClient } from '../client';
import type { ConformanceResponse, DiagnosticsResponse } from '../types';

export interface ConformanceCheckOptions {
  datasetId: string;
  modelId: string;
  method?: 'token_replay' | 'alignment';
}

export interface ConformanceResult {
  id: string;
  datasetId: string;
  modelId: string;
  fitness: number;
  precision?: number;
  generalization?: number;
  simplicity?: number;
  method: string;
  isConformant: boolean;
  fittingTraces: number;
  totalTraces: number;
  createdAt: string;
}

export interface ConformanceModule {
  check: (options: ConformanceCheckOptions) => Promise<ConformanceResult>;
  getDiagnostics: (datasetId: string, modelId: string) => Promise<DiagnosticsResponse>;
}

function transformConformanceResponse(be: ConformanceResponse): ConformanceResult {
  return {
    id: be.id,
    datasetId: be.dataset_id,
    modelId: be.model_id,
    fitness: be.fitness,
    precision: be.precision,
    generalization: be.generalization,
    simplicity: be.simplicity,
    method: be.method,
    isConformant: be.is_conformant,
    fittingTraces: be.fitting_traces,
    totalTraces: be.total_traces,
    createdAt: be.created_at,
  };
}

export function createConformanceModule(client: ApiClient): ConformanceModule {
  return {
    async check(options: ConformanceCheckOptions) {
      const response = await client.post<ConformanceResponse>('/conformance/check', {
        dataset_id: options.datasetId,
        model_id: options.modelId,
        method: options.method ?? 'token_replay',
      });
      return transformConformanceResponse(response);
    },

    async getDiagnostics(datasetId: string, modelId: string) {
      return client.get<DiagnosticsResponse>(
        `/conformance/diagnostics/${encodeURIComponent(datasetId)}/${encodeURIComponent(modelId)}`
      );
    },
  };
}

File: backend/src/features/process_mining/schemas/conformance.py
"""Conformance schemas - Conformance checking and alignment.

Contains schemas for:
- Conformance check requests and responses
- Deviation and diagnostics
- Alignment moves and case alignments
- Quality metrics
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict

from src.features.process_mining.enums import ConformanceMethod
from src.shared.schemas import PaginatedResponse


class ConformanceCheckRequest(BaseModel):
    """Request for conformance checking."""

    dataset_id: str
    model_id: str
    method: ConformanceMethod = ConformanceMethod.TOKEN_REPLAY


class ConformanceResponse(BaseModel):
    """Conformance check result."""

    id: str
    dataset_id: str
    model_id: str
    fitness: float
    precision: float | None
    generalization: float | None = None  # FE expects this metric
    simplicity: float | None = None  # FE expects this metric
    method: str  # What method was requested
    algorithm_used: str  # What algorithm actually ran (may differ if fallback)
    fallback_reason: str | None = None  # Reason for fallback if different from method
    is_conformant: bool  # fitness >= 0.8
    fitting_traces: int
    total_traces: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class DeviationDetail(BaseModel):
    """Structured deviation for conformance diagnostics."""

    case_id: str
    activity: str
    violation_type: str  # 'missing_activity' | 'wrong_order' | 'extra_activity'
    expected_after: str | None = None
    frequency: int = 1
    impact: str = "medium"  # 'low' | 'medium' | 'high'


class DiagnosticsResponse(BaseModel):
    """Detailed conformance diagnostics."""

    fitness: float
    precision: float | None
    generalization: float | None = None
    simplicity: float | None = None
    f_score: float | None = None  # Harmonic mean of fitness & precision
    total_traces: int
    fitting_traces: int
    non_fitting_traces: int
    fitness_ratio: float
    average_alignment_cost: float | None = None
    deviations: list[DeviationDetail] | None = None  # Structured deviations


class QualityMetricsResponse(BaseModel):
    """Full quality metrics for a process model.

    Contains all 4 quality dimensions from PM4py:
    - Fitness: How well the log fits the model
    - Precision: How much the model allows for behavior not in the log
    - Generalization: How well the model generalizes beyond observed behavior
    - Simplicity: How simple/understandable the model is
    """

    dataset_id: str
    model_id: str
    fitness: float
    precision: float | None = None
    generalization: float | None = None
    simplicity: float | None = None
    f_score: float | None = None  # Harmonic mean of fitness & precision


class ConformanceListResponse(PaginatedResponse):
    """Paginated conformance results list."""

    items: list[ConformanceResponse]


class DeviationResponse(BaseModel):
    """Conformance deviation detail."""

    case_id: str
    activity: str | None
    deviation_type: str
    details: str


class AlignmentMove(BaseModel):
    """Single move in an alignment sequence."""

    log_move: str | None = None
    model_move: str | None = None
    move_type: str  # 'sync', 'log_only', 'model_only'


class CaseAlignmentResponse(BaseModel):
    """Alignment result for a single case."""

    case_id: str
    fitness: float
    cost: int = 0
    alignment: list[AlignmentMove]
    is_fit: bool = True


class AlignmentDiagnosticsResponse(BaseModel):
    """Response for alignment diagnostics endpoint."""

    dataset_id: str
    model_id: str
    total_cases: int
    fitting_cases: int
    average_fitness: float
    case_alignments: list[CaseAlignmentResponse]
